Mage/src/main/java/mage/abilities/effects/common/GainsChoiceOfAbilitiesEffect.java

```diff
@@ -0,0 +1,120 @@
+import mage.choices.ChoiceImpl;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.targetpointer.FixedTargets;
+
+import java.util.*;
+
+/**
+ * Something gains one ability of the player's choice from a list of abilities until end of turn.
+ * Affects any targets set and optionally the source as well. Have no targets to affect only the source.
+ *
+ * @author notgreat, TheElk801
+ */
+public class GainsChoiceOfAbilitiesEffect extends OneShotEffect {
+
+    private final Map<String, Ability> abilityMap = new LinkedHashMap<>();
+    private final boolean affectSource, includeEnd;
+    private final String targetDescription;
+
+    public GainsChoiceOfAbilitiesEffect(Ability... abilities) {
+        this(false, null, true, abilities);
+    }
+
+    public GainsChoiceOfAbilitiesEffect(boolean affectSource, Ability... abilities) {
+        this(affectSource, null, true, abilities);
+    }
+
+    public GainsChoiceOfAbilitiesEffect(boolean affectSource, String targetDescription, boolean includeEnd, Ability... abilities) {
+        super(Outcome.AddAbility);
+        this.affectSource = affectSource;
+        this.targetDescription = targetDescription;
+        this.includeEnd = includeEnd;
+        for (Ability ability : abilities) {
+            this.abilityMap.put(ability.getRule(), ability);
+        }
+    }
+
+    protected GainsChoiceOfAbilitiesEffect(final GainsChoiceOfAbilitiesEffect effect) {
+        super(effect);
+        this.affectSource = effect.affectSource;
+        this.abilityMap.putAll(effect.abilityMap);
+        this.targetDescription = effect.targetDescription;
+        this.includeEnd = effect.includeEnd;
+    }
+
+    @Override
+    public GainsChoiceOfAbilitiesEffect copy() {
+        return new GainsChoiceOfAbilitiesEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        List<Permanent> permanents = new ArrayList<>();
+        if (affectSource) {
+            permanents.add(source.getSourcePermanentIfItStillExists(game));
+        }
+        if (getTargetPointer().isInitialized()) {
+            for (UUID p : getTargetPointer().getTargets(game, source)) {
+                permanents.add(game.getPermanent(p));
+            }
+        }
+        permanents.removeIf(Objects::isNull);
+        if (player == null || permanents.isEmpty()) {
+            return false;
+        }
+        Choice choice = new ChoiceImpl(true);
+        choice.setMessage("Choose an ability");
+        choice.setChoices(new HashSet<>(abilityMap.keySet()));
+        player.choose(outcome, choice, game);
+        Ability ability = abilityMap.getOrDefault(choice.getChoice(), null);
+        if (ability != null) {
+            game.addEffect(new GainAbilityTargetEffect(ability, Duration.EndOfTurn)
```

Comment:
For info: I don’t look at all cards here but if something use existing ability to gain (from real another object) then it must be added with additional param (that it was copied from existing object).